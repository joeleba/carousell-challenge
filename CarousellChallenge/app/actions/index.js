export const ADD_POST = "ADD_POST";
export const ADD_COMMENT = "ADD_COMMENT";
export const UPVOTE = "UPVOTE";
export const DOWNVOTE = "DOWNVOTE";

let nextContentId = 0
export const addPost = (title, text) => ({
  type: ADD_POST,
  id: nextContentId++,
  title: title,
  text
})

export const addComment = (parent_id, text) => ({
  type: ADD_COMMENT,
  id: nextContentId++,
  parent_id: parent_id,
  text
})

export const upvote = (id) => ({
  type: UPVOTE,
  id
})

export const downvote = (id) => ({
  type: DOWNVOTE,
  id
})
