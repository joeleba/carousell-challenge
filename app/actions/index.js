export const ADD_POST = 'ADD_POST';
export const ADD_COMMENT = 'ADD_COMMENT';
export const UPVOTE = 'UPVOTE';
export const DOWNVOTE = 'DOWNVOTE';
export const POST_TYPE = 'POST';
export const COMMENT_TYPE = 'COMMENT'

export const addPost = (title, text) => ({
  type: ADD_POST,
  title,
  text,
});

export const addComment = (parentId, parentType, text) => ({
  type: ADD_COMMENT,
  parentId,
  parentType,
  text,
});

export const upvote = (id, contentType) => ({
  type: UPVOTE,
  contentType,
  id,
});

export const downvote = (id, contentType) => ({
  type: DOWNVOTE,
  contentType,
  id,
});
