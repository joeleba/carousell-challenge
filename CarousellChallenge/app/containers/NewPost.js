import React, { Component } from 'react'
import { AppRegistry, Text, TextInput, Button, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux'
import { NavigationActions } from 'react-navigation';
import Icon from 'react-native-vector-icons/FontAwesome';

import { addPost } from '../actions'
import NewContentForm from '../components/NewContentForm'

// TODO: Refactor to Component
class NewPost extends Component {
  render() {
    let form;
    let title;
    return (
      <View style={styles.container}>
        <Icon
          name="close"
          size={20}
          onPress={() => this.props.dispatchNavigateBackToPostListView()}>
        </Icon>

        <TextInput
          onChangeText={(text) => { title = text }}
          style={styles.titleInput}
          placeholder="Title"
        />
        <NewContentForm ref={node => { form = node }}/>
        <Button
            onPress={() => {
              let newPostId = this.props.currentId;
              this.props.dispatchAddPost(title, form.state.content);
              this.props.dispatchNavigateToPost(newPostId);
            }}
            title="Submit"
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
   flex: 1,
   margin: 20
  },
  titleInput: {
    marginBottom: 20,
    marginTop: 20
  },
})

function mapStateToProps (state) {
  return {
    currentId: state.contentReducers.currentId
  }
}

function mapDispatchToProps (dispatch) {
  return {
    dispatchNavigateToPost: (postId) =>
      dispatch(NavigationActions.navigate({
        routeName: 'PostView',
        params: {
          postId: postId
        }
      })),
    dispatchNavigateBackToPostListView: () =>
      dispatch(NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'PostListView' })],
      })),
    dispatchAddPost: (title, content) =>
      dispatch(addPost(title, content))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(NewPost)
