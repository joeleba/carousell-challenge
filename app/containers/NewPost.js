import React, { Component } from 'react';
import { TextInput, Button, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import Icon from '@expo/vector-icons/FontAwesome';

import { addPost } from '../actions';
import NewContentForm from '../components/NewContentForm';

class NewPost extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formState: {
        content: '',
      },
      title: '',
    };
  }

  render() {
    return (
      <View style={styles.container}>
        <Icon
          name="close"
          size={20}
          onPress={() => this.props.dispatchNavigateBackToPostListView()}
        />

        <TextInput
          onChangeText={(text) => { this.state.title = text; }}
          style={styles.titleInput}
          placeholder="Title"
        />
        <NewContentForm onChange={(formState) => {
          this.setState({ formState });
        }} />

        <Button
          onPress={() => {
            let newPostId = this.props.currentPostId;
            this.props.dispatchAddPost(this.state.title, this.state.formState.content);
            this.props.dispatchNavigateToPost(newPostId);
          }}
          title="Submit"
          disabled={this.state.formState.content.length === 0 || this.state.title.length === 0}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 20,
  },
  titleInput: {
    marginBottom: 20,
    marginTop: 20,
  },
});

function mapStateToProps(state) {
  return {
    currentPostId: state.contentReducers.currentPostId,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToPost: postId =>
      dispatch(NavigationActions.navigate({
        routeName: 'PostView',
        params: { postId },
      })),
    dispatchNavigateBackToPostListView: () =>
      dispatch(NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'PostListView' })],
      })),
    dispatchAddPost: (title, content) => dispatch(addPost(title, content)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(NewPost);
