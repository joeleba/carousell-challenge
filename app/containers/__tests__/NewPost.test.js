import React from 'react';
import { Text, FlatList, StyleSheet, View } from 'react-native';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import NewPost from '../NewPost';
import { addComment } from '../../actions';

const mockStore = configureStore();

const stateNoPostNoComment = {
  contentReducers: {
    currentPostId: 0,
    posts: {},
    comments: {},
  },
};

const navigationProp = {
  state: {
    params: {
      replyingToObject: {
        content: 'abc',
        id: 0
      },
      postId: 0
    }
  }
};

describe('Component: NewPost', () => {
  it('renders with nested comments', () => {
    const store = mockStore(stateNoPostNoComment);
    const wrapper = shallow(
      <NewPost />,
      { context: { store: store } },
    );
    const render = wrapper.dive();
    expect(render).toMatchSnapshot();
    expect(wrapper.prop('currentPostId')).toEqual(0);
  });

  it('dispatches correct actions on submit', () => {
    const store = mockStore(stateNoPostNoComment);
    const wrapper = shallow(
      <NewPost />,
      { context: { store: store } },
    );
    const render = wrapper.dive();
    render.find('Button').first().simulate('press');

    const actions = store.getActions();
    const expectedPayload = [
      { type: 'ADD_POST', title: '', text: '' },
      {"params": {"postId": 0}, "routeName": "PostView", "type": "Navigation/NAVIGATE"}
    ];

    expect(actions).toEqual(expectedPayload);
  });

  it('dispatches correct actions on back', () => {
    const store = mockStore(stateNoPostNoComment);
    const wrapper = shallow(
      <NewPost />,
      { context: { store: store } },
    );
    const render = wrapper.dive();
    render.find('[name="close"]').first().simulate('press');

    const actions = store.getActions();
    const expectedPayload = [
      {
        "actions": [
          {
            "routeName": "PostListView",
            "type": "Navigation/NAVIGATE"
          }
        ],
        "index": 0,
        "key": undefined,
        "type": "Navigation/RESET"
      }
    ];

    expect(actions).toEqual(expectedPayload);
  });
});
