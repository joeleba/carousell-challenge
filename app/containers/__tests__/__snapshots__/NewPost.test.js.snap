// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NewPost renders with nested comments 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 20,
    }
  }
>
  <Icon
    allowFontScaling={false}
    name="close"
    onPress={[Function]}
    size={20}
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Title"
    style={
      Object {
        "marginBottom": 20,
        "marginTop": 20,
      }
    }
  />
  <NewContentForm
    onChange={[Function]}
  />
  <Button
    disabled={true}
    onPress={[Function]}
    title="Submit"
  />
</View>
`;
